Inform extensions of user/pass changes

ktsakalozos opened the pull request.

freeekanayaka commented:
Unit tests where broken in master, but I just fixed that in #12 . Mind rebasing your branch against master? That will fix the unit tests.

Also, please add an integration test to this branch, proving that the change you propose actually works as intended. I think the test should be added to tests/external.py, but most probably you'll need to change that test file a bit, since it currently exercises the jenkins:extension interface using the ci-configurator charm, which won't be able to trigger the code path that is being proposed in this branch. Ideally this branch should get rid of the ci-configurator branch and introduce a local test charm implementing the jenkins:extension interface and providing ways to introspect the state (e.g. store the jenkins password that it gets through the relation into a file), so the integration tests can perform assertions conveniently.

ryan-beisner commented:
There should be corresponding unit and amulet test coverage for the new config option. This reactive rewrite has pretty good unit and amulet test coverage and I'd really like to see us keep that level of trust as contributions are made. Thank you.

ktsakalozos commented:
I will try to work on this today.

@ryan-beisner +1 for having tests. Allow me to just clarify that this PR is not a new feature, it is a bug fix. If you change the password of Jenkins all related extensions need to know about the new credentials so as to keep functioning properly.

ktsakalozos commented:
In the latest commit you will find:
1) An amulet test that uses the cwr charm (probably the only charm that uses the extension-interface. That test is very similar to the external test.
2) Removed the trusty tests. They are failing since this charm is a xenial only series
3) Minor lint fixes and increase test timeouts (tests were failing on my local lxd env behind a 50Mbps line)

- Other Amulet tests are failing as well and i am going to open issues for those.
- My suggestion would be to have some bundles to run integration tests. Cwr-ci could be one.

freeekanayaka commented:
Konstantinos Tsakalozos <notifications@github.com> writes:

> In the latest commit you will find:

> 1) An amulet test that uses the cwr charm (probably the only charm
>    that uses the extension-interface. That test is very similar to the
>    external test.

As a propbably mentioned somewhere else, my personal preference would be
to have a minimal test charm in the tree (e.g. under tests/) that
implements the extension-interface and do performs some dummy
side-effects with it that can be inspected by the test code
(e.g. writing files on the unit etc).

Since I understand this is a somehow sophisticated approach (that
probably no other other or very few charms have taken so far), I
wouldn't block on it, but I'm mentioning it as something to keep in mind
medium term (I'll probably work on it at some point).

> 2) Removed the trusty tests. They are failing since this charm is a
>    xenial only series

I'm pretty sure they used to pass on trusty, so probably something
outside this charge has changed/regressed. If folks have no interest in
keeping trusty support, I'm fine dropping it.

[...]

- My suggestion would be to have some bundles to run
> integration tests. Cwr-ci could be one.

Here my guts is -1 unless persuaded otherwise.

If there's a bundle, integration tests should be in that bundle (and
typically exercise what's specific to that bundle, without repeating
charm-level integration tests).

I would avoid any situations where:

- The integration tests for the jenkins charm are coupled with some
  specific bundle.

- The integration tests of the jenkins charm are coupled with some
  specific charm that consumes the interfaces that the jenkins charm
  provides (this is already the case in the current master where
  integration tests depend on the ci-configurator charm to exercise
  the 'external' interface).

All this would contribute in making the tests more brittle and harder to
debug: unrelated/buggy/regressing changes in the third-party bundle or
charm would make the integration tests of the jenkins charm fail for no
good reason (the bug is not in the jenkins charm).

Plus, it typically makes tests more expensive to run, because the
third-party bundle or charm will inevitably slow down things (since they
will execute some real beast as opposed to some slim test code, like the
dummy test charm I described above).

It's better to place these tests in the third-party bundles or charms
that consume this lower-level charm (jenkins).

Tests are much more valuable when they run quickly and they point
quickly to where problems are.

If we keep the jenkins charm tests free from any coupling with third
party bundles/charms, the authors of those third party bundles/charm can
know with a good degree of confidence where problems are (e.g. if the
low-level jenkins charm integration tests are green, but their own tests
are failing, the problem is most certaintly in the third-party code).

ktsakalozos commented:
I think that as soon as we added the series in metadata we dropped support for trusty.

https://github.com/jenkinsci/jenkins-charm/commit/4e86635080c5e72c0ccf259834c94182d4df3edd

ktsakalozos commented:
We mostly agree about the integration tests. The way the amulet tests are structured right now we essentially deploy bundles (we deploy hsproxy, ci-configurator, a slave).

freeekanayaka commented:
Konstantinos Tsakalozos <notifications@github.com> writes:

> We mostly agree about the integration tests. The way the amulet tests
> are structured right now we essentially deploy bundles (we deploy
> hsproxy, ci-configurator, a slave).

So, as I mentioned the use of ci-configurator in the current tests is
subobtimal for the reasons stated in my other reply.

On the contrary, I think that the use of haproxy is a bit more
legitimate because we are consumers of the interfaces it provides (in
some way we depend on it and it doesn't depend on us).

I would work towards dropping the suboptimal cases (e.g. the use of
ci-configurator), rather than adding new ones (e.g. cwr or cwr-ci).

When I get some time, I'll give a spin at dropping ci-configurator.

If you want to add a case that involves the actual cwr charm, because
you don't want to create a testable dummy charm that implements the
jenkins-extension interface, that's fine. But keep in mind that the
cwr(-ci) work should not rely on those tests in any way, since when I'll
drop the use of ci-configurator I would drop other similar cases like
cwr.

If you want to have integration coverage for jenkins+cwr, please add
integration tests in the cwr charm and/or cwr-ci bundle, exactly in the
same way we use haproxy in our tests but haproxy doesn't use jenkins in
its tests (which would be equivalent as using cwr in jenkins tessts).

freeekanayaka commented on tests/extension.py:
```diff
@@ -0,0 +1,49 @@
+        self.deployment.add("cwr", EXTERNAL[self.series])
+        self.deployment.relate(
+            "jenkins:extension", "cwr:jenkins")
+        self.deployment.expose("cwr")
```
Why do you need to expose the cwr application?

johnsca replied:
The `ExtensionDeploymentTest` connects to it (on port 5000) to confirm that it can interact properly with Jenkins.

freeekanayaka commented:
Cory Johns <notifications@github.com> writes:

> johnsca commented on this pull request.
>
>
>
>> +import time
> +from deployment import DeploymentTest
> +from basic import BasicDeploymentSpec
> +
> +EXTERNAL = {
> +    "xenial": "cs:~juju-solutions/cwr-11",
> +}
> +
> +
> +class ExtensionDeploymentSpec(BasicDeploymentSpec):
> +
> +    def _pre_setup_00_extension(self):
> +        self.deployment.add("cwr", EXTERNAL[self.series])
> +        self.deployment.relate(
> +            "jenkins:extension", "cwr:jenkins")
> +        self.deployment.expose("cwr")
>
> The `ExtensionDeploymentTest` connects to it (on port 5000) to confirm that it can interact properly with Jenkins.

Gotcha. Didn't realize that cwr exposed it's own web UI.
